modules/ethereum/src/verification.rs

```diff
@@ -31,39 +31,205 @@
 // along with Parity-Bridge.  If not, see <http://www.gnu.org/licenses/>.
 
 use crate::error::Error;
-use crate::validators::step_validator;
-use crate::{AuraConfiguration, ImportContext, Storage};
-use primitives::{public_to_address, Address, Header, SealedEmptyStep, H256, H520, U128, U256};
+use crate::validators::{step_validator, Validators, ValidatorsConfiguration};
+use crate::{AuraConfiguration, ImportContext, PoolConfiguration, ScheduledChange, Storage};
+use codec::Encode;
+use primitives::{public_to_address, Address, Header, Receipt, SealedEmptyStep, H256, H520, U128, U256};
 use sp_io::crypto::secp256k1_ecdsa_recover;
+use sp_std::{vec, vec::Vec};
+
+/// Pre-check to see if should try and import this header.
+/// Returns error if we should not try to import this block.
+/// Returns hash of the header and number of the last finalized block otherwise.
+pub fn is_importable_header<S: Storage>(storage: &S, header: &Header) -> Result<(H256, H256), Error> {
+	// we never import any header that competes with finalized header
+	let (finalized_block_number, finalized_block_hash) = storage.finalized_block();
+	if header.number <= finalized_block_number {
+		return Err(Error::AncientHeader);
+	}
+	// we never import any header with known hash
+	let hash = header.hash();
+	if storage.header(&hash).is_some() {
+		return Err(Error::KnownHeader);
+	}
+
+	Ok((hash, finalized_block_hash))
+}
+
+/// Try accept unsigned aura header into transaction pool.
+pub fn accept_aura_header_into_pool<S: Storage>(
+	storage: &S,
+	config: &AuraConfiguration,
+	validators_config: &ValidatorsConfiguration,
+	pool_config: &PoolConfiguration,
+	header: &Header,
+	receipts: Option<&Vec<Receipt>>,
+) -> Result<(Vec<Vec<u8>>, Vec<Vec<u8>>), Error> {
+	// check if we can verify further
+	let (hash, _) = is_importable_header(storage, header)?;
+
+	// we can always do contextless checks
+	contextless_checks(config, header)?;
+
+	// we want to avoid having same headers twice in the pool
+	// => we're strict about receipts here - if we need them, we require receipts to be Some,
+	// otherwise we require receipts to be None
+	let receipts_required = Validators::new(validators_config).maybe_signals_validators_change(header);
+	match (receipts_required, receipts.is_some()) {
+		(true, false) => return Err(Error::MissingTransactionsReceipts),
+		(false, true) => return Err(Error::RedundantTransactionsReceipts),
+		_ => (),
+	}
+
+	// we do not want to have all future headers in the pool at once
+	// => if we see header with number > maximal ever seen header number + LIMIT,
+	// => we consider this transaction invalid, but only at this moment (we do not want to ban it)
+	// => let's mark it as Unknown transaction
+	let (best_number, best_hash, _) = storage.best_block();
+	let difference = header.number.saturating_sub(best_number);
+	if difference > pool_config.max_future_number_difference {
```

Comment:
We should use `difference` to prioritise transactions in the pool. I.e. the ones with higher difference should have priority bump.

The resoning behind it would be, that if we ever have multiple bridges included in one runtime, we want the bridges that are the most behind to be processed earlier.

Reply:
I don't think it'll work in that way. So you have two bridges. The first knows about block#100 and chain tip is at block#1000. The second knows about block#200 and tip is at block#205. So we are submitting tx to bridge#1 with block#101 and to bridge#2 with block #201. Difference is 1 for both transactions => we can't prioritize bridge#1 over bridge#2.

I think what you're talking about is more fits in #38 - after we'll make something like this, we will be able to determine how far beyond tip we are and prioritize accordingly.